app/src/main/java/de/westnordost/streetcomplete/data/StreetCompleteDatabase.kt

```diff
@@ -0,0 +1,278 @@
+        } else if (oldVersion < newVersion) {
+            DatabaseInitializer.onUpgrade(this, oldVersion, newVersion)
+            databaseConnection.execSQL("PRAGMA user_version = $newVersion")
+        }
+    }
+
+    override fun exec(sql: String, args: Array<Any>?) {
+        databaseConnection.prepare(sql).use { statement ->
+            statement.bindAll(args)
+            statement.step()
+        }
+    }
+
+    override fun <T> rawQuery(
+        sql: String,
+        args: Array<Any>?,
+        transform: (CursorPosition) -> T,
+    ): List<T> =
+        databaseConnection.prepare(sql).use { statement ->
+            statement.bindAll(args)
+            statement.toSequence(transform).toList()
+        }
+
+    override fun <T> queryOne(
+        table: String,
+        columns: Array<String>?,
+        where: String?,
+        args: Array<Any>?,
+        groupBy: String?,
+        having: String?,
+        orderBy: String?,
+        transform: (CursorPosition) -> T,
+    ): T? =
+        databaseConnection.prepareQuery(false, table, columns, where, groupBy, having, orderBy, 1).use { statement ->
+            statement.bindAll(args)
+            statement.toSequence(transform).firstOrNull()
+        }
+
+    override fun <T> query(
+        table: String,
+        columns: Array<String>?,
+        where: String?,
+        args: Array<Any>?,
+        groupBy: String?,
+        having: String?,
+        orderBy: String?,
+        limit: Int?,
+        distinct: Boolean,
+        transform: (CursorPosition) -> T,
+    ): List<T> =
+        databaseConnection.prepareQuery(distinct, table, columns, where, groupBy, having, orderBy, limit).use { statement ->
+            statement.bindAll(args)
+            statement.toSequence(transform).toList()
+        }
+
+    override fun insert(
+        table: String,
+        values: Collection<Pair<String, Any?>>,
+        conflictAlgorithm: ConflictAlgorithm?,
+    ): Long =
+        databaseConnection.prepareInsert(table, values.map { it.first }, conflictAlgorithm).use { statement ->
+            statement.bindAll(values.map { it.second }.toTypedArray())
+            statement.executeInsert()
+        }
+
+    override fun insertMany(
+        table: String,
+        columnNames: Array<String>,
+        valuesList: Iterable<Array<Any?>>,
+        conflictAlgorithm: ConflictAlgorithm?,
+    ): List<Long> {
+        val statement = databaseConnection.prepareInsert(table, columnNames.toList(), conflictAlgorithm)
+        val result = ArrayList<Long>()
+        transaction {
+            for (values in valuesList) {
+                require(values.size == columnNames.size)
+                statement.bindAll(values)
+                val rowId = statement.executeInsert()
+                result.add(rowId)
+                statement.clearBindings()
+                statement.reset()
+            }
+            statement.close()
+        }
+        return result
```

Comment:
By the way, this can potentially be optimized (quite a bit?).

SQLite allows a syntax like this:
```SQL
INSERT INTO my_table (column_a, column_b, column_c)
VALUES (1,2,"a"), (2,4,"b"), (3,1,"c"), (4,5,"d")
```
to insert 4 values. I.e. no need for a transaction, no need for repeating the first line x times.

To be honest, I thought I've already long done this, but apparently not.

https://sqlite.org/lang_insert.html

Reply:
(But before blindly doing a premature optimization, it might be worth it to confirm that this would actually be faster in some unit test or something. Maybe binding like tens of thousands of parameters to a prepared statement with tens of thousands placeholders comes with implications or limitations in itself. Also, it needn't be part of this PR.)